resources/src/asset.rs

```diff
@@ -0,0 +1,117 @@
+use utils;
+/// Permissions around who can own an asset and whether or
+/// not the asset issuer can freeze the asset.
+#[derive(Serialize, Deserialize, Debug)]
+struct Flag {
+    auth_required: bool,
+    auth_revocable: bool,
+}
+
+/// Assets are the units that are traded on the Stellar Network.
+/// An asset consists of an type, code, and issuer.
+/// Any asset can be traded for any other asset.
+/// https://www.stellar.org/developers/horizon/reference/resources/asset.html
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Asset {
+    asset_type: String,
+    asset_code: String,
+    asset_issuer: String,
+    #[serde(deserialize_with = "utils::deserialize_from_str")] amount: String,
+    formatted_amount: Option<i64>,
+    num_accounts: u32,
+    flags: Flag,
+}
+
+impl Asset {
+    /// The type of this asset: “credit_alphanum4”, or “credit_alphanum12”.
+    /// Returns a slice that lives as long as the asset does.
+    pub fn asset_type<'a>(&'a self) -> &'a str {
+        &self.asset_type
+    }
+
+    /// The code of this asset.
+    /// Returns a slice that lives as long as the asset does.
+    pub fn asset_code<'a>(&'a self) -> &'a str {
+        &self.asset_code
+    }
+
+    /// The issuer of this asset.  This corresponds to the id of an account.
+    /// Returns a slice that lives as long as the asset does.
+    pub fn asset_issuer<'a>(&'a self) -> &'a str {
+        &self.asset_issuer
+    }
+
+    /// The number of units of credit issued for this asset.
+    /// This number is scaled by 10 million to display the number if the format a
+    /// user would expect it in
+    /// https://www.stellar.org/developers/guides/concepts/assets.html
+    /// Returns a signed 64-bit integer.
+    pub fn amount(&self) -> i64 {
+        match self.formatted_amount {
+            Some(formatted_amount) => return formatted_amount,
+            None => {
+                // A string with 1 char has len() == 1
+                let length = self.amount.len();
+                // A string starting with . has rfind('.') == 0
+                let decimal_place = self.amount.rfind('.').unwrap();
+                // Extra -1 to account for indexing differences
+                let number_decimals = length - 1 - decimal_place;
+                let mut new_amount = self.amount.clone();
+                // remove the decimal
+                let new_amount = new_amount.replace(".", "");
+                // Stellar sends a float that is reduced from true value by 10^7 so raise by 10 
+                // minus the amount we gained from removing decimal
+                let required_power: u32 = (7 - number_decimals) as u32;
+                return (new_amount.parse::<i64>().unwrap()) * 10_i64.pow(required_power);
```

Comment:
I would love to find a way to set the optional value within this None arm, but couldn't find anything and going to be.  @kbacha maybe you've got a nice idea of how to handle this.

Reply:
You cannot mutate a non-mut self reference. We should tackle the deserialize before the struct is made.